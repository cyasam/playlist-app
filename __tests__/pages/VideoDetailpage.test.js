import React from 'react'
import { shallow } from 'enzyme'
import { VideoDetailPage, mapStateToProps } from '../../src/scripts/pages/VideoDetailpage'

describe('Video Detail Page', () => {
  const setup = (propOverrides) => {
    const props = {
      history: {
        listen: jest.fn(),
        push: jest.fn(),
        location: { pathname: '/watch', search: `?v=yG-9XIIELk4` }
      },
      videoDetail: {
        isFetching: false,
        error: null,
        video: {
          id: 'lZoA5ZX4wC0',
          title: 'Ocean Beach Surfing Raw | Sanâ€¦cisco, CA',
          channelTitle: 'Jeff Chavolla',
          viewCount: '121034',
          publishedAt: '2017-12-04T07:18:50.000Z',
          description: 'Raw surfing video clips of Ocean Beach in San Francisco',
          thumbnail: {
            url: 'https://i.ytimg.com/vi/lZoA5ZX4wC0/hqdefault.jpg',
            width: 480,
            height: 360
          }
        }
      },
      fetchVideoDetail: jest.fn(),
      trendings: {
        isFetching: false,
        videos: [
          {
            id: 'lZoA5ZX4wC0',
            title: 'Video Title',
            description: 'Lorem ipsum dolor sit amed.',
            thubmnail: 'video-thumbnail.jpg'
          },
          {
            id: 'lZoA5ZX4wC01',
            title: 'Video Title 2',
            description: 'Lorem ipsum dolor sit amed.',
            thubmnail: 'video-thumbnail-2.jpg'
          }
        ]
      },
      fetchTrendings: jest.fn(),
      ...propOverrides
    }

    const wrapper = shallow(<VideoDetailPage { ...props } />)
    wrapper.instance().historyListener = jest.fn()

    return {
      wrapper,
      props
    }
  }

  it('renders properly', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  it('calls fetchVideoDetail action when first page load', () => {
    const { wrapper, props } = setup({ videoDetail: { isFetching: true, video: {} } })

    wrapper.instance().componentDidMount()
    expect(wrapper).toMatchSnapshot()
  })

  it('returns videoDetail object when initialize component', () => {
    const { wrapper, props } = setup()
    expect(mapStateToProps(props).videoDetail).toEqual(props.videoDetail)
  })

  it('calls fetchData when calling history listen', () => {
  
  })

  it('remove history listen when calling WillUnmount', () => {
    const { wrapper } = setup()
    wrapper.instance().componentWillUnmount()

    const mockHistoryListener = jest.spyOn(wrapper.instance(), 'historyListener')
    expect(mockHistoryListener).toHaveBeenCalled()
  })

  it('pushs history to Homepage action when video id is not defined in query string', () => {
    const { wrapper, props } = setup({ 
      history: {
        listen: jest.fn(),
        push: jest.fn(),
        location: {
          pathname: '/watch',
          search: '?v='
        }
      },
      videoDetail: { isFetching: true, video: {} }
    })

    wrapper.instance().componentDidMount()
    expect(props.history.push).toHaveBeenCalled()
  })

  it('returns nothing if the page is not video detail', () => {
    const { wrapper, props } = setup({ 
      history: {
        listen: jest.fn(),
        push: jest.fn(),
        location: {
          pathname: '/'
        }
      }
    })

    expect(wrapper.instance().fetchData()).toEqual(null)
  })
})