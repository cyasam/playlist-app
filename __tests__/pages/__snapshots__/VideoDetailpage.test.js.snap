// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video Detail Page calls fetchVideoDetail action when first page load 1`] = `
<SideEffect(DocumentTitle)
  title="Playlist App"
>
  <div
    className="detail-page"
  >
    <div
      className="row"
    >
      <div
        className="video-detail-box col-8"
      >
        <VideoDetail
          data={
            Object {
              "isFetching": true,
              "video": Object {},
            }
          }
        />
      </div>
      <div
        className="list-box col-4"
      >
        <VideoList
          isFetching={false}
          listType="media"
          videos={
            Array [
              Object {
                "description": "Lorem ipsum dolor sit amed.",
                "id": "lZoA5ZX4wC0",
                "thubmnail": "video-thumbnail.jpg",
                "title": "Video Title",
              },
              Object {
                "description": "Lorem ipsum dolor sit amed.",
                "id": "lZoA5ZX4wC01",
                "thubmnail": "video-thumbnail-2.jpg",
                "title": "Video Title 2",
              },
            ]
          }
        />
      </div>
    </div>
  </div>
</SideEffect(DocumentTitle)>
`;

exports[`Video Detail Page calls fetchVideoDetail when page changed. 1`] = `
<SideEffect(DocumentTitle)
  title="Ocean Beach Surfing Raw | San…cisco, CA - Playlist App"
>
  <div
    className="detail-page"
  >
    <div
      className="row"
    >
      <div
        className="video-detail-box col-8"
      >
        <VideoDetail
          data={
            Object {
              "error": null,
              "isFetching": false,
              "video": Object {
                "channelTitle": "Jeff Chavolla",
                "description": "Raw surfing video clips of Ocean Beach in San Francisco",
                "id": "lZoA5ZX4wC0",
                "publishedAt": "2017-12-04T07:18:50.000Z",
                "thumbnail": Object {
                  "height": 360,
                  "url": "https://i.ytimg.com/vi/lZoA5ZX4wC0/hqdefault.jpg",
                  "width": 480,
                },
                "title": "Ocean Beach Surfing Raw | San…cisco, CA",
                "viewCount": "121034",
              },
            }
          }
        />
      </div>
      <div
        className="list-box col-4"
      >
        <VideoList
          isFetching={false}
          listType="media"
          videos={
            Array [
              Object {
                "description": "Lorem ipsum dolor sit amed.",
                "id": "lZoA5ZX4wC0",
                "thubmnail": "video-thumbnail.jpg",
                "title": "Video Title",
              },
              Object {
                "description": "Lorem ipsum dolor sit amed.",
                "id": "lZoA5ZX4wC01",
                "thubmnail": "video-thumbnail-2.jpg",
                "title": "Video Title 2",
              },
            ]
          }
        />
      </div>
    </div>
  </div>
</SideEffect(DocumentTitle)>
`;

exports[`Video Detail Page calls fetchVideoDetail when page doesn't changed. 1`] = `
<SideEffect(DocumentTitle)
  title="Ocean Beach Surfing Raw | San…cisco, CA - Playlist App"
>
  <div
    className="detail-page"
  >
    <div
      className="row"
    >
      <div
        className="video-detail-box col-8"
      >
        <VideoDetail
          data={
            Object {
              "error": null,
              "isFetching": false,
              "video": Object {
                "channelTitle": "Jeff Chavolla",
                "description": "Raw surfing video clips of Ocean Beach in San Francisco",
                "id": "lZoA5ZX4wC0",
                "publishedAt": "2017-12-04T07:18:50.000Z",
                "thumbnail": Object {
                  "height": 360,
                  "url": "https://i.ytimg.com/vi/lZoA5ZX4wC0/hqdefault.jpg",
                  "width": 480,
                },
                "title": "Ocean Beach Surfing Raw | San…cisco, CA",
                "viewCount": "121034",
              },
            }
          }
        />
      </div>
      <div
        className="list-box col-4"
      >
        <VideoList
          isFetching={false}
          listType="media"
          videos={
            Array [
              Object {
                "description": "Lorem ipsum dolor sit amed.",
                "id": "lZoA5ZX4wC0",
                "thubmnail": "video-thumbnail.jpg",
                "title": "Video Title",
              },
              Object {
                "description": "Lorem ipsum dolor sit amed.",
                "id": "lZoA5ZX4wC01",
                "thubmnail": "video-thumbnail-2.jpg",
                "title": "Video Title 2",
              },
            ]
          }
        />
      </div>
    </div>
  </div>
</SideEffect(DocumentTitle)>
`;

exports[`Video Detail Page renders properly 1`] = `
<SideEffect(DocumentTitle)
  title="Ocean Beach Surfing Raw | San…cisco, CA - Playlist App"
>
  <div
    className="detail-page"
  >
    <div
      className="row"
    >
      <div
        className="video-detail-box col-8"
      >
        <VideoDetail
          data={
            Object {
              "error": null,
              "isFetching": false,
              "video": Object {
                "channelTitle": "Jeff Chavolla",
                "description": "Raw surfing video clips of Ocean Beach in San Francisco",
                "id": "lZoA5ZX4wC0",
                "publishedAt": "2017-12-04T07:18:50.000Z",
                "thumbnail": Object {
                  "height": 360,
                  "url": "https://i.ytimg.com/vi/lZoA5ZX4wC0/hqdefault.jpg",
                  "width": 480,
                },
                "title": "Ocean Beach Surfing Raw | San…cisco, CA",
                "viewCount": "121034",
              },
            }
          }
        />
      </div>
      <div
        className="list-box col-4"
      >
        <VideoList
          isFetching={false}
          listType="media"
          videos={
            Array [
              Object {
                "description": "Lorem ipsum dolor sit amed.",
                "id": "lZoA5ZX4wC0",
                "thubmnail": "video-thumbnail.jpg",
                "title": "Video Title",
              },
              Object {
                "description": "Lorem ipsum dolor sit amed.",
                "id": "lZoA5ZX4wC01",
                "thubmnail": "video-thumbnail-2.jpg",
                "title": "Video Title 2",
              },
            ]
          }
        />
      </div>
    </div>
  </div>
</SideEffect(DocumentTitle)>
`;
